libp2p/stream/lpstream.nim

```diff
@@ -94,19 +94,26 @@ method closed*(s: LPStream): bool {.base, inline.} =
 method atEof*(s: LPStream): bool {.base, inline.} =
   s.isEof
 
-method readExactly*(s: LPStream,
-                    pbytes: pointer,
-                    nbytes: int):
-                    Future[void] {.base, async.} =
-  doAssert(false, "not implemented!")
-
 method readOnce*(s: LPStream,
                  pbytes: pointer,
                  nbytes: int):
                  Future[int]
   {.base, async.} =
   doAssert(false, "not implemented!")
 
+proc readExactly*(s: LPStream,
+                    pbytes: pointer,
+                    nbytes: int):
+                    Future[void] {.async.} =
+
+  if s.atEof:
+    raise newLPStreamEOFError()
+
+  var pbuffer = cast[ptr UncheckedArray[byte]](pbytes)
+  var read = 0
+  while read < nbytes and not(s.atEof()):
```

Comment:
if it reads less than nbytes, it should notify the caller (ie then it's not exactly any more) - ie when `atEof` is true.

Reply:
(through an exception most likely)

Reply:
yeah, good point.